runtime/commonMain/src/kotlinx/serialization/json/JsonElementSerializer.kt

```diff
@@ -67,8 +67,9 @@ public object JsonPrimitiveSerializer : KSerializer<JsonPrimitive> {
 
     override fun deserialize(decoder: Decoder): JsonPrimitive {
         verify(decoder)
-        return if (decoder.decodeNotNullMark()) JsonPrimitive(decoder.decodeString())
-        else decoder.decodeSerializableValue(JsonNullSerializer)
+        val result = (decoder as JsonInput).decodeJson()
+        if (result !is JsonPrimitive) throw JsonDecodingException(-1, "Unexpected JSON element, expected JsonPrimitive, had ${result::class}", result.toString())
```

Comment:
and `result.primitive` ext val. It does not contain `result.toString()` currently, but we can add this (still we should mind that `result.toString()` can be rather big JSON)